File: tema_2/frontend/src/services/admin.service.js
// src/services/admin.service.js
import axios from '../utils/axiosConfig';

const AdminService = {
  // User (Receptionist) Management
  createReceptionist: async (userData) => {
    const response = await axios.post('/admin/users', userData);
    return response.data;
  },
  
  getAllReceptionists: async () => {
    const response = await axios.get('/admin/users');
    return response.data;
  },
  
  // Doctor Management
  createDoctor: async (doctorData) => {
    const response = await axios.post('/admin/doctors', doctorData);
    return response.data;
  },
  
  updateDoctor: async (id, doctorData) => {
    const response = await axios.put(`/admin/doctors/${id}`, doctorData);
    return response.data;
  },
  
  deleteDoctor: async (id) => {
    const response = await axios.delete(`/admin/doctors/${id}`);
    return response.data;
  },
  
  getAllDoctors: async () => {
    const response = await axios.get('/admin/doctors');
    return response.data;
  },
  
  // Medical Service Management
  createService: async (serviceData) => {
    const response = await axios.post('/admin/services', serviceData);
    return response.data;
  },
  
  updateService: async (id, serviceData) => {
    const response = await axios.put(`/admin/services/${id}`, serviceData);
    return response.data;
  },
  
  deleteService: async (id) => {
    const response = await axios.delete(`/admin/services/${id}`);
    return response.data;
  },
  
  getAllServices: async () => {
    const response = await axios.get('/admin/services');
    return response.data;
  },
  
  // Format date to ISO_DATE_TIME
  formatDateToISODateTime: (dateStr) => {
    // Ensure we have a valid date string
    if (!dateStr) return '';
    
    // If it's already in ISO format with time, return as is
    if (dateStr.includes('T')) return dateStr;
    
    // Otherwise add time component
    return `${dateStr}T00:00:00`;
  },
  
  // Reports
  getAppointmentsReport: async (startDate, endDate) => {
    const formattedStartDate = AdminService.formatDateToISODateTime(startDate);
    // For the end date, use end of day
    const formattedEndDate = endDate ? `${endDate}T23:59:59` : '';
    
    const response = await axios.get(`/admin/reports/appointments?startDate=${formattedStartDate}&endDate=${formattedEndDate}`);
    return response.data.appointments || [];
  },
  
  getDoctorsReport: async (startDate, endDate) => {
    const formattedStartDate = AdminService.formatDateToISODateTime(startDate);
    // For the end date, use end of day
    const formattedEndDate = endDate ? `${endDate}T23:59:59` : '';
    
    const response = await axios.get(`/admin/reports/doctors?startDate=${formattedStartDate}&endDate=${formattedEndDate}`);
    return response.data.doctorStats || [];
  },
  
  getServicesReport: async (startDate, endDate) => {
    const formattedStartDate = AdminService.formatDateToISODateTime(startDate);
    // For the end date, use end of day
    const formattedEndDate = endDate ? `${endDate}T23:59:59` : '';
    
    const response = await axios.get(`/admin/reports/services?startDate=${formattedStartDate}&endDate=${formattedEndDate}`);
    return response.data.serviceStats || [];
  }
};

export default AdminService;

File: tema_2/backend/src/main/java/com/medcare/service/AdminService.java
package com.medcare.service;

import com.medcare.model.*;
import com.medcare.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class AdminService {

    private final UserService userService;
    private final DoctorRepository doctorRepository;
    private final MedicalServiceRepository medicalServiceRepository;
    private final AppointmentRepository appointmentRepository;

    public AdminService(UserService userService, DoctorRepository doctorRepository,
                       MedicalServiceRepository medicalServiceRepository,
                       AppointmentRepository appointmentRepository) {
        this.userService = userService;
        this.doctorRepository = doctorRepository;
        this.medicalServiceRepository = medicalServiceRepository;
        this.appointmentRepository = appointmentRepository;
    }

    @Transactional
    public User createReceptionist(User user) {
        user.setRole(User.UserRole.RECEPTIONIST);
        return userService.createUser(user);
    }

    public List<User> getAllReceptionists() {
        return userService.getAllReceptionists();
    }

    @Transactional
    public Doctor createDoctor(Doctor doctor) {
        return doctorRepository.save(doctor);
    }

    @Transactional
    public Doctor updateDoctor(Long id, Doctor doctor) {
        Doctor existingDoctor = doctorRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Doctor not found"));
        
        existingDoctor.setName(doctor.getName());
        existingDoctor.setSpecialization(doctor.getSpecialization());
        existingDoctor.setStartTime(doctor.getStartTime());
        existingDoctor.setEndTime(doctor.getEndTime());
        
        return doctorRepository.save(existingDoctor);
    }

    @Transactional
    public void deleteDoctor(Long id) {
        Doctor doctor = doctorRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Doctor not found"));
        doctor.setActive(false);
        doctorRepository.save(doctor);
    }

    public List<Doctor> getAllDoctors() {
        return doctorRepository.findByActiveTrue();
    }

    @Transactional
    public MedicalService createService(MedicalService service) {
        return medicalServiceRepository.save(service);
    }

    @Transactional
    public MedicalService updateService(Long id, MedicalService service) {
        MedicalService existingService = medicalServiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Service not found"));
        
        existingService.setName(service.getName());
        existingService.setPrice(service.getPrice());
        existingService.setDuration(service.getDuration());
        
        return medicalServiceRepository.save(existingService);
    }

    @Transactional
    public void deleteService(Long id) {
        MedicalService service = medicalServiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Service not found"));
        service.setActive(false);
        medicalServiceRepository.save(service);
    }

    public List<MedicalService> getAllServices() {
        return medicalServiceRepository.findByActiveTrue();
    }

    public Map<String, Object> getAppointmentsReport(String startDate, String endDate) {
        LocalDateTime start = parseDateTime(startDate);
        LocalDateTime end = parseDateTime(endDate);
        
        List<Appointment> appointments = appointmentRepository.findAppointmentsBetweenDates(start, end);
        
        return Map.of(
            "totalAppointments", appointments.size(),
            "appointments", appointments
        );
    }

    public Map<String, Object> getDoctorsReport(String startDate, String endDate) {
        LocalDateTime start = parseDateTime(startDate);
        LocalDateTime end = parseDateTime(endDate);
        
        List<Object[]> doctorStats = appointmentRepository.findMostRequestedDoctors(start, end);
        
        return Map.of(
            "doctorStats", doctorStats.stream()
                .map(stat -> Map.of(
                    "doctor", stat[0],
                    "appointmentCount", stat[1]
                ))
                .collect(Collectors.toList())
        );
    }

    public Map<String, Object> getServicesReport(String startDate, String endDate) {
        LocalDateTime start = parseDateTime(startDate);
        LocalDateTime end = parseDateTime(endDate);
        
        List<Object[]> serviceStats = appointmentRepository.findMostRequestedServices(start, end);
        
        return Map.of(
            "serviceStats", serviceStats.stream()
                .map(stat -> Map.of(
                    "service", stat[0],
                    "appointmentCount", stat[1]
                ))
                .collect(Collectors.toList())
        );
    }

    private LocalDateTime parseDateTime(String dateTimeStr) {
        return LocalDateTime.parse(dateTimeStr, DateTimeFormatter.ISO_DATE_TIME);
    }
} 